File: OnaPlotter/Utilities/Format.cs
namespace OnaPlotter.Utilities;

public static class Format
{
    /// <summary>Conversion factor: meters per second to knots.</summary>
    public const double MsToKnots = 1.94384;

    /// <summary>Conversion factor: radians to degrees.</summary>
    public const double RadToDeg = 180.0 / Math.PI;

    /// <summary>Meters per nautical mile.</summary>
    public const double MetersPerNm = 1852.0;

    /// <summary>Formats a speed value in m/s as knots, one decimal. Returns "--" for null.</summary>
    public static string Speed(double? ms) =>
        ms is null ? "--" : (ms.Value * MsToKnots).ToString("F1");

    /// <summary>Formats an angle in radians as degrees (0-360), no decimals. Returns "--" for null.</summary>
    public static string Degrees(double? rad)
    {
        if (rad is null) return "--";
        double deg = rad.Value * RadToDeg;
        if (deg < 0) deg += 360;
        return deg.ToString("F0");
    }

    /// <summary>Formats a distance in meters as nautical miles (2 decimals under 10nm, 1 above).</summary>
    public static string Nm(double? meters)
    {
        if (meters is null) return "--";
        double nm = meters.Value / MetersPerNm;
        return nm < 10 ? nm.ToString("F2") : nm.ToString("F1");
    }

    /// <summary>Formats a distance in meters: meters under 1000, nautical miles above.</summary>
    public static string MetersDual(double? m)
    {
        if (m is null) return "--";
        return m.Value < 1000 ? $"{m.Value:F0}m" : $"{(m.Value / MetersPerNm):F2}nm";
    }

    /// <summary>Formats a depth in meters, one decimal. Returns "--" for null.</summary>
    public static string Depth(double? m) =>
        m is null ? "--" : m.Value.ToString("F1");

    /// <summary>Formats remaining time in seconds as "1h23m" or "5m30s".</summary>
    public static string TimeToGo(double? seconds)
    {
        if (seconds is null) return "--";
        var ts = TimeSpan.FromSeconds(seconds.Value);
        return ts.TotalHours >= 1
            ? $"{(int)ts.TotalHours}h{ts.Minutes:D2}m"
            : $"{ts.Minutes}m{ts.Seconds:D2}s";
    }

    /// <summary>Formats ETA as "HH:mm" today, "dd.MM HH:mm" otherwise.</summary>
    public static string Eta(double? ttgSeconds)
    {
        if (ttgSeconds is null) return "--";
        var eta = DateTime.Now.AddSeconds(ttgSeconds.Value);
        return eta.Date == DateTime.Today ? eta.ToString("HH:mm") : eta.ToString("dd.MM HH:mm");
    }

    /// <summary>Formats elapsed-since-MOB-raise as "T+5s" / "T+12m"
    /// / "T+1h23m". Pure formatter - caller passes the seconds
    /// since the casualty was raised. JS layer uses this via the
    /// mirrored helper in <c>format.js</c>; tests live here to pin
    /// the exact strings.</summary>
    public static string MobElapsed(int seconds)
    {
        if (seconds < 0) seconds = 0;
        if (seconds < 60) return $"T+{seconds}s";
        int min = seconds / 60;
        if (min < 60) return $"T+{min}m";
        int hr = min / 60;
        return $"T+{hr}h{min % 60}m";
    }

    /// <summary>Formats a lat/lon pair as DMS with hemisphere
    /// indicators: "47.50000&deg;N 8.50000&deg;W". Five decimals on
    /// the degrees ~= 1m precision - enough for a chart popup. Used
    /// by waypoint / note / MOB popups in JS via the mirrored helper.</summary>
    public static string LatLonDms(double lat, double lon)
    {
        char ns = lat >= 0 ? 'N' : 'S';
        char ew = lon >= 0 ? 'E' : 'W';
        return $"{Math.Abs(lat).ToString("F5", System.Globalization.CultureInfo.InvariantCulture)}°{ns} " +
               $"{Math.Abs(lon).ToString("F5", System.Globalization.CultureInfo.InvariantCulture)}°{ew}";
    }

    /// <summary>Formats a radius in nautical miles for an on-chart
    /// ring label: "0.5 nm" / "1.5 nm" / "5 nm". Two decimals below
    /// 1 nm so 0.5 doesn't round to "1"; one decimal at 1-9.99;
    /// whole number above 10. Trailing zeros are stripped ("0.50"
    /// -&gt; "0.5", "1.0" -&gt; "1") to keep chart labels compact.
    /// Returns empty for non-finite or non-positive input.</summary>
    public static string RangeRingLabel(double nm)
    {
        if (!double.IsFinite(nm) || nm <= 0) return string.Empty;
        var inv = System.Globalization.CultureInfo.InvariantCulture;
        string digits;
        if (nm < 1)
        {
            // F2 then strip trailing zeros (and the dot if all zeros after).
            digits = nm.ToString("F2", inv).TrimEnd('0').TrimEnd('.');
        }
        else if (nm < 10)
        {
            // F1 then strip a trailing ".0" - a non-zero tenths digit
            // (e.g. "1.5") survives the strip.
            digits = nm.ToString("F1", inv);
            if (digits.EndsWith(".0", StringComparison.Ordinal))
            {
                digits = digits[..^2];
            }
        }
        else
        {
            // AwayFromZero matches JS's Math.round (50.5 -> 51) so the
            // C# canonical and the format.js mirror produce identical
            // strings on the .5 boundary.
            digits = ((long)Math.Round(nm, MidpointRounding.AwayFromZero)).ToString(inv);
        }
        return $"{digits} nm";
    }

    /// <summary>Formats XTE with port/starboard suffix. SignalK convention: positive = starboard.</summary>
    public static string Xte(double? meters)
    {
        if (meters is null) return "--";
        double abs = Math.Abs(meters.Value);
        string side = meters > 0 ? " S" : meters < 0 ? " P" : "";
        return abs < 1000 ? $"{abs:F0}m{side}" : $"{(abs / MetersPerNm):F2}nm{side}";
    }

    /// <summary>XTE severity class name for CSS color coding.</summary>
    public static string XteClass(double? meters) => meters switch
    {
        null => "",
        var x when Math.Abs(x.Value) < 50 => "xte-ok",
        var x when Math.Abs(x.Value) < 200 => "xte-warn",
        _ => "xte-danger"
    };

    /// <summary>Depth severity class name for CSS color coding.</summary>
    public static string DepthClass(double? depth) => depth switch
    {
        null => "",
        < 3 => "depth-danger",
        < 8 => "depth-warn",
        _ => "depth-ok"
    };
}

File: OnaPlotter/wwwroot/js/format.js
// Single source of truth for the JS-side renderer formatters.
//
// Every helper here mirrors a tested C# canonical in
// OnaPlotter/Utilities/ - the C# tests pin the exact strings, this
// module reproduces the formula so JS-tick-rate callers don't need
// an interop round-trip per render. If you change one side, change
// the other - the C# tests are the contract; this file is the
// renderer-side mirror.
//
// Cross-references (each function names its C# canonical):
//   - msToKnots      -> Format.MsToKnots                (Format.cs)
//   - speedColor     -> SpeedColor.Rgb                  (SpeedColor.cs)
//   - speedBucket    -> SpeedColor.Bucket               (SpeedColor.cs)
//   - stalenessOp    -> StalenessOpacity.Compute        (StalenessOpacity.cs)
//   - mobElapsed     -> Format.MobElapsed               (Format.cs)
//   - latLonDms      -> Format.LatLonDms                (Format.cs)
//   - rangeRingLabel -> Format.RangeRingLabel           (Format.cs)
//   - etaWithTtg     -> Format.EtaWithTtg               (Format.cs)

// --- Constants (mirror Format.cs) -------------------------------

/** m/s -> knots multiplier. Matches Format.MsToKnots. */
export const MS_TO_KNOTS = 1.94384;

/** SpeedColor.DefaultRgb - default colour when SOG is unknown. */
export const SPEED_COLOR_DEFAULT = '#3b82f6';

/** SpeedColor.Buckets - per-bucket lower bound (m/s). */
export const SPEED_BUCKETS = [0, 1, 2, 3, 5, 8];

/** StalenessOpacity threshold constants. */
export const STALE_FRESH_SECONDS = 30;
export const STALE_FADED_SECONDS = 300;
export const STALE_FLOOR_OPACITY = 0.25;
export const STALE_FADE_SPAN = 0.65;

// --- Helpers ----------------------------------------------------

/**
 * SpeedColor.Rgb mirror. Returns "rgb(r,g,b)" for the given SOG
 * (m/s), or SPEED_COLOR_DEFAULT when sog is null/undefined.
 * Two-stage lerp: 0..3 kn blue->green; 3..8 kn green->yellow.
 */
export function speedColor(sogMs) {
    if (sogMs == null) return SPEED_COLOR_DEFAULT;
    const kn = sogMs * MS_TO_KNOTS;
    const t = Math.min(kn / 8, 1);
    if (t < 0.5) {
        const f = t * 2;
        const r = Math.round(59 + f * (34 - 59));
        const g = Math.round(130 + f * (197 - 130));
        const b = Math.round(246 + f * (94 - 246));
        return `rgb(${r},${g},${b})`;
    } else {
        const f = (t - 0.5) * 2;
        const r = Math.round(34 + f * (234 - 34));
        const g = Math.round(197 + f * (179 - 197));
        const b = Math.round(94 + f * (8 - 94));
        return `rgb(${r},${g},${b})`;
    }
}

/** SpeedColor.Bucket mirror. Returns the bucket index (0..5). */
export function speedBucket(sogMs) {
    if (sogMs == null) return 0;
    for (let i = SPEED_BUCKETS.length - 1; i >= 0; i--) {
        if (sogMs >= SPEED_BUCKETS[i]) return i;
    }
    return 0;
}

/**
 * StalenessOpacity.Compute mirror. Returns the inline-style opacity
 * string for an AIS marker of the given age (seconds), or null when
 * the marker is fresh (caller clears any previous inline opacity so
 * CSS defaults apply).
 */
export function stalenessOpacity(ageSec) {
    if (ageSec < STALE_FRESH_SECONDS) return null;
    if (ageSec >= STALE_FADED_SECONDS) {
        return STALE_FLOOR_OPACITY.toFixed(2);
    }
    const op = 1 - STALE_FADE_SPAN * (ageSec - STALE_FRESH_SECONDS) / (STALE_FADED_SECONDS - STALE_FRESH_SECONDS);
    return op.toFixed(2);
}

/**
 * Format.MobElapsed mirror. Returns "T+5s" / "T+12m" / "T+1h23m"
 * for elapsed seconds since the MOB raise. Negative input clamps
 * to "T+0s".
 */
export function mobElapsed(seconds) {
    let sec = Math.floor(seconds);
    if (sec < 0) sec = 0;
    if (sec < 60) return `T+${sec}s`;
    const min = Math.floor(sec / 60);
    if (min < 60) return `T+${min}m`;
    const hr = Math.floor(min / 60);
    return `T+${hr}h${min % 60}m`;
}

/** Single-axis lat formatter - "47.50000°N". Used by popup
 * tables that put lat / lon in separate cells. */
export function latDms(lat) {
    const ns = lat >= 0 ? 'N' : 'S';
    return `${Math.abs(lat).toFixed(5)}°${ns}`;
}

/** Single-axis lon formatter - "8.50000°E". */
export function lonDms(lon) {
    const ew = lon >= 0 ? 'E' : 'W';
    return `${Math.abs(lon).toFixed(5)}°${ew}`;
}

/**
 * Format.LatLonDms mirror. Returns "47.50000°N 8.50000°W"
 * (or with a custom separator e.g. ", ") with five decimals
 * (~1m precision). Default separator is a single space, matching
 * the C# canonical - callers that need a comma pass
 * <c>latLonDms(lat, lon, ', ')</c>.
 */
export function latLonDms(lat, lon, sep = ' ') {
    return `${latDms(lat)}${sep}${lonDms(lon)}`;
}

/**
 * Format.RangeRingLabel mirror. Returns "0.5 nm" / "1.5 nm" /
 * "5 nm" / "" for non-finite or zero-or-negative input. Trailing
 * zeros are stripped ("0.50" -> "0.5", "1.0" -> "1") to keep
 * chart labels compact - both AIS guard rings and radar range
 * rings use this formatter.
 */
export function rangeRingLabel(nm) {
    if (!isFinite(nm) || nm <= 0) return '';
    let digits;
    if (nm < 1) {
        digits = nm.toFixed(2).replace(/\.?0+$/, '');
    } else if (nm < 10) {
        digits = nm.toFixed(1).replace(/\.0$/, '');
    } else {
        digits = Math.round(nm).toString();
    }
    return `${digits} nm`;
}

/**
 * RouteEta.Format mirror - canonical: OnaPlotter/Utilities/RouteEta.cs.
 * Returns the popup ETA line "ETA HH:MM (in 1h 23m)" / "ETA HH:MM
 * (in 5m)" / "ETA HH:MM (in >99h)", or null on null/non-finite/non-
 * positive input (callers drop the row in those cases). Hour cap
 * (99h), minute round, 1m floor all match the C# port; the parity
 * test scrapes this file for the matching tokens (see
 * RouteEtaJsParityTests).
 */
export function etaWithTtg(ttgSeconds) {
    if (ttgSeconds == null || !isFinite(ttgSeconds) || ttgSeconds <= 0) return null;
    const totalMin = Math.max(1, Math.round(ttgSeconds / 60));
    const totalH = Math.floor(totalMin / 60);
    let inText;
    if (totalH > 99) {
        inText = '>99h';
    } else if (totalH > 0) {
        inText = `${totalH}h ${totalMin % 60}m`;
    } else {
        inText = `${totalMin}m`;
    }
    const eta = new Date(Date.now() + ttgSeconds * 1000);
    const pad2 = (n) => String(n).padStart(2, '0');
    const clock = `${pad2(eta.getHours())}:${pad2(eta.getMinutes())}`;
    return `ETA ${clock} (in ${inText})`;
}
